File: truss-compiler/src/main/java/truss/compiler/syntax/SyntaxNode.java
package truss.compiler.syntax;

import truss.compiler.Span;
import org.apache.commons.lang.Validate;

import java.util.Collections;
import java.util.List;

public abstract class SyntaxNode {
    private final Span span;

    protected SyntaxNode(Span span) {
        Validate.notNull(span, "span");

        this.span = span;
    }

    public abstract SyntaxKind getKind();

    public Span getSpan() {
        return span;
    }

    public abstract void accept(SyntaxVisitor visitor) throws Exception;

    public abstract <T> T accept(SyntaxActionVisitor<T> visitor) throws Exception;

    public List<? extends SyntaxNode> getChildren() {
        return Collections.emptyList();
    }
}

File: Truss.Compiler/Syntax/SyntaxNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Truss.Compiler.Syntax {
    public abstract class SyntaxNode {
        protected SyntaxNode(Span span) {
            if (span == null) {
                throw new ArgumentNullException("span");
            }

            Span = span;
        }

        public abstract SyntaxKind Kind { get; }

        public Span Span { get; private set; }

        public abstract void Accept(ISyntaxVisitor visitor);

        public abstract T Accept<T>(ISyntaxVisitor<T> visitor);
    }
}
